From 022e90093ad121008af1ebfe8aa134363f1878db Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 8 Mar 2022 17:09:59 +0100
Subject: [PATCH 029/954] refactor(st-ddr): gather redundant procedures as
 generic services

Disabling AXI port, enabling host interface and both enabling/disabling
software self-refresh services are already present inside the driver
source code.
Factorize by gathering them as services inside the generic part,
and adapt driver to call these new functions.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a08d33ff57226c6eed89aec21c27eeed615b1ed3 refactor(st-ddr): gather redundant procedures as generic services]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I880f88b1cb6fc76199ad9ea33e9d63a5c469aed4
---
 drivers/st/ddr/stm32mp1_ddr.c         | 11 +---
 drivers/st/ddr/stm32mp1_ddr_helpers.c | 70 ++++------------------
 drivers/st/ddr/stm32mp_ddr.c          | 83 +++++++++++++++++++++++++++
 include/drivers/st/stm32mp_ddr.h      |  4 ++
 4 files changed, 100 insertions(+), 68 deletions(-)

diff --git a/drivers/st/ddr/stm32mp1_ddr.c b/drivers/st/ddr/stm32mp1_ddr.c
index 1cbc55582e..1259ddf0c5 100644
--- a/drivers/st/ddr/stm32mp1_ddr.c
+++ b/drivers/st/ddr/stm32mp1_ddr.c
@@ -508,8 +508,7 @@ static void stm32mp1_ddr3_dll_off(struct stm32mp_ddr_priv *priv)
 #endif
 
 	/* 12. Exit the self-refresh state by setting PWRCTL.selfref_sw = 0. */
-	mmio_clrbits_32((uintptr_t)&priv->ctl->pwrctl,
-			DDRCTRL_PWRCTL_SELFREF_SW);
+	stm32mp_ddr_sw_selfref_exit(priv->ctl);
 	stm32mp1_wait_operating_mode(priv, DDRCTRL_STAT_OPERATING_MODE_NORMAL);
 
 	/*
@@ -524,10 +523,7 @@ static void stm32mp1_ddr3_dll_off(struct stm32mp_ddr_priv *priv)
 	 */
 
 	/* 15. Write DBG1.dis_hif = 0 to re-enable reads and writes. */
-	mmio_clrbits_32((uintptr_t)&priv->ctl->dbg1, DDRCTRL_DBG1_DIS_HIF);
-	VERBOSE("[0x%lx] dbg1 = 0x%x\n",
-		(uintptr_t)&priv->ctl->dbg1,
-		mmio_read_32((uintptr_t)&priv->ctl->dbg1));
+	stm32mp_ddr_enable_host_interface(priv->ctl);
 }
 
 static void stm32mp1_refresh_disable(struct stm32mp_ddrctl *ctl)
@@ -839,8 +835,7 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 
 	/* Trigger self-refresh exit */
 	if (config->self_refresh) {
-		mmio_clrbits_32((uintptr_t)&priv->ctl->pwrctl,
-				DDRCTRL_PWRCTL_SELFREF_SW);
+		stm32mp_ddr_sw_selfref_exit(priv->ctl);
 	}
 
 	stm32mp1_wait_operating_mode(priv, DDRCTRL_STAT_OPERATING_MODE_NORMAL);
diff --git a/drivers/st/ddr/stm32mp1_ddr_helpers.c b/drivers/st/ddr/stm32mp1_ddr_helpers.c
index 4808221986..9bf5c1c8cf 100644
--- a/drivers/st/ddr/stm32mp1_ddr_helpers.c
+++ b/drivers/st/ddr/stm32mp1_ddr_helpers.c
@@ -34,11 +34,6 @@ void ddr_enable_clock(void)
 
 static int ddr_sw_self_refresh_in(void)
 {
-	uint64_t timeout;
-	uint32_t stat;
-	uint32_t operating_mode;
-	uint32_t selref_type;
-	uint8_t op_mode_changed = 0;
 	uintptr_t rcc_base = stm32mp_rcc_base();
 	uintptr_t pwr_base = stm32mp_pwr_base();
 	uintptr_t ddrctrl_base = stm32mp_ddrctrl_base();
@@ -51,47 +46,14 @@ static int ddr_sw_self_refresh_in(void)
 	stm32mp1_clk_rcc_regs_unlock();
 
 	/* Blocks AXI ports from taking anymore transactions */
-	mmio_clrbits_32(ddrctrl_base + DDRCTRL_PCTRL_0,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#if STM32MP_DDR_DUAL_AXI_PORT
-	mmio_clrbits_32(ddrctrl_base + DDRCTRL_PCTRL_1,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#endif
-
-	/* Waits unit all AXI ports are idle
-	 * Poll PSTAT.rd_port_busy_n = 0
-	 * Poll PSTAT.wr_port_busy_n = 0
-	 */
-	timeout = timeout_init_us(TIMEOUT_500US);
-	while (mmio_read_32(ddrctrl_base + DDRCTRL_PSTAT)) {
-		if (timeout_elapsed(timeout)) {
-			goto pstat_failed;
-		}
+	if (stm32mp_ddr_disable_axi_port((struct stm32mp_ddrctl *)ddrctrl_base) != 0U) {
+		goto pstat_failed;
 	}
-	/* SW Self-Refresh entry */
-	mmio_setbits_32(ddrctrl_base + DDRCTRL_PWRCTL,
-			DDRCTRL_PWRCTL_SELFREF_SW);
 
-	/* Wait operating mode change in self-refresh mode
-	 * with STAT.operating_mode[1:0]==11.
-	 * Ensure transition to self-refresh was due to software
-	 * by checking also that STAT.selfref_type[1:0]=2.
-	 */
-	timeout = timeout_init_us(TIMEOUT_500US);
-	while (!timeout_elapsed(timeout)) {
-		stat = mmio_read_32(ddrctrl_base + DDRCTRL_STAT);
-		operating_mode = stat & DDRCTRL_STAT_OPERATING_MODE_MASK;
-		selref_type = stat & DDRCTRL_STAT_SELFREF_TYPE_MASK;
-
-		if ((operating_mode == DDRCTRL_STAT_OPERATING_MODE_SR) &&
-		    (selref_type == DDRCTRL_STAT_SELFREF_TYPE_SR)) {
-			op_mode_changed = 1;
-			break;
-		}
-	}
-
-	if (op_mode_changed == 0U)
+	/* SW Self-Refresh entry */
+	if (stm32mp_ddr_sw_selfref_entry((struct stm32mp_ddrctl *)ddrctrl_base) != 0U) {
 		goto selfref_sw_failed;
+	}
 
 	/* IOs powering down (PUBL registers) */
 	mmio_setbits_32(ddrphyc_base + DDRPHYC_ACIOCR, DDRPHYC_ACIOCR_ACPDD);
@@ -183,17 +145,11 @@ static int ddr_sw_self_refresh_in(void)
 	return 0;
 
 selfref_sw_failed:
-	/* This bit should be cleared to restore DDR in its previous state */
-	mmio_clrbits_32(ddrctrl_base + DDRCTRL_PWRCTL,
-			DDRCTRL_PWRCTL_SELFREF_SW);
+	/* Restore DDR in its previous state */
+	stm32mp_ddr_sw_selfref_exit((struct stm32mp_ddrctl *)ddrctrl_base);
 
 pstat_failed:
-	mmio_setbits_32(ddrctrl_base + DDRCTRL_PCTRL_0,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#if STM32MP_DDR_DUAL_AXI_PORT
-	mmio_setbits_32(ddrctrl_base + DDRCTRL_PCTRL_1,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#endif
+	stm32mp_ddr_enable_axi_port((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	return -1;
 }
@@ -318,8 +274,7 @@ int ddr_sw_self_refresh_exit(void)
 			DDRPHYC_DSGCR_CKEPDD_MASK);
 
 	/* Remove selfrefresh */
-	mmio_clrbits_32(ddrctrl_base + DDRCTRL_PWRCTL,
-			DDRCTRL_PWRCTL_SELFREF_SW);
+	stm32mp_ddr_sw_selfref_exit((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	/* Wait operating_mode == normal */
 	timeout = timeout_init_us(TIMEOUT_500US);
@@ -332,12 +287,7 @@ int ddr_sw_self_refresh_exit(void)
 	}
 
 	/* AXI ports are no longer blocked from taking transactions */
-	mmio_setbits_32(ddrctrl_base + DDRCTRL_PCTRL_0,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#if STM32MP_DDR_DUAL_AXI_PORT
-	mmio_setbits_32(ddrctrl_base + DDRCTRL_PCTRL_1,
-			DDRCTRL_PCTRL_N_PORT_EN);
-#endif
+	stm32mp_ddr_enable_axi_port((struct stm32mp_ddrctl *)ddrctrl_base);
 
 	stm32mp1_clk_rcc_regs_lock();
 
diff --git a/drivers/st/ddr/stm32mp_ddr.c b/drivers/st/ddr/stm32mp_ddr.c
index 6776e3ba8c..2f085ac80e 100644
--- a/drivers/st/ddr/stm32mp_ddr.c
+++ b/drivers/st/ddr/stm32mp_ddr.c
@@ -93,7 +93,90 @@ void stm32mp_ddr_enable_axi_port(struct stm32mp_ddrctl *ctl)
 	VERBOSE("[0x%lx] pctrl_1 = 0x%x\n", (uintptr_t)&ctl->pctrl_1,
 		mmio_read_32((uintptr_t)&ctl->pctrl_1));
 #endif
+}
+
+int stm32mp_ddr_disable_axi_port(struct stm32mp_ddrctl *ctl)
+{
+	uint64_t timeout;
+	uint32_t pstat;
+
+	/* Disable uMCTL2 AXI port 0 */
+	mmio_clrbits_32((uintptr_t)&ctl->pctrl_0, DDRCTRL_PCTRL_N_PORT_EN);
+	VERBOSE("[0x%lx] pctrl_0 = 0x%x\n", (uintptr_t)&ctl->pctrl_0,
+		mmio_read_32((uintptr_t)&ctl->pctrl_0));
+
+#if STM32MP_DDR_DUAL_AXI_PORT
+	/* Disable uMCTL2 AXI port 1 */
+	mmio_clrbits_32((uintptr_t)&ctl->pctrl_1, DDRCTRL_PCTRL_N_PORT_EN);
+	VERBOSE("[0x%lx] pctrl_1 = 0x%x\n", (uintptr_t)&ctl->pctrl_1,
+		mmio_read_32((uintptr_t)&ctl->pctrl_1));
+#endif
 
+	/*
+	 * Waits until all AXI ports are idle
+	 * Poll PSTAT.rd_port_busy_n = 0
+	 * Poll PSTAT.wr_port_busy_n = 0
+	 */
+	timeout = timeout_init_us(TIMEOUT_US_1S);
+	do {
+		pstat = mmio_read_32((uintptr_t)&ctl->pstat);
+		VERBOSE("[0x%lx] pstat = 0x%x ",
+			(uintptr_t)&ctl->pstat, pstat);
+		if (timeout_elapsed(timeout)) {
+			return -1;
+		}
+	} while (pstat != 0U);
+
+	return 0;
+}
+
+void stm32mp_ddr_enable_host_interface(struct stm32mp_ddrctl *ctl)
+{
+	mmio_clrbits_32((uintptr_t)&ctl->dbg1, DDRCTRL_DBG1_DIS_HIF);
+	VERBOSE("[0x%lx] dbg1 = 0x%x\n",
+		(uintptr_t)&ctl->dbg1,
+		mmio_read_32((uintptr_t)&ctl->dbg1));
+}
+
+int stm32mp_ddr_sw_selfref_entry(struct stm32mp_ddrctl *ctl)
+{
+	uint64_t timeout;
+	uint32_t stat;
+	uint32_t operating_mode;
+	uint32_t selref_type;
+
+	mmio_setbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_SELFREF_SW);
+	VERBOSE("[0x%lx] pwrctl = 0x%x\n",
+		(uintptr_t)&ctl->pwrctl,
+		mmio_read_32((uintptr_t)&ctl->pwrctl));
+
+	/*
+	 * Wait operating mode change in self-refresh mode
+	 * with STAT.operating_mode[1:0]==11.
+	 * Ensure transition to self-refresh was due to software
+	 * by checking also that STAT.selfref_type[1:0]=2.
+	 */
+	timeout = timeout_init_us(TIMEOUT_500US);
+	while (!timeout_elapsed(timeout)) {
+		stat = mmio_read_32((uintptr_t)&ctl->stat);
+		operating_mode = stat & DDRCTRL_STAT_OPERATING_MODE_MASK;
+		selref_type = stat & DDRCTRL_STAT_SELFREF_TYPE_MASK;
+
+		if ((operating_mode == DDRCTRL_STAT_OPERATING_MODE_SR) &&
+		    (selref_type == DDRCTRL_STAT_SELFREF_TYPE_SR)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+void stm32mp_ddr_sw_selfref_exit(struct stm32mp_ddrctl *ctl)
+{
+	mmio_clrbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_SELFREF_SW);
+	VERBOSE("[0x%lx] pwrctl = 0x%x\n",
+		(uintptr_t)&ctl->pwrctl,
+		mmio_read_32((uintptr_t)&ctl->pwrctl));
 }
 
 int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
diff --git a/include/drivers/st/stm32mp_ddr.h b/include/drivers/st/stm32mp_ddr.h
index 4ec0dae431..bfaaa52ac6 100644
--- a/include/drivers/st/stm32mp_ddr.h
+++ b/include/drivers/st/stm32mp_ddr.h
@@ -65,6 +65,10 @@ void stm32mp_ddr_set_reg(const struct stm32mp_ddr_priv *priv, enum stm32mp_ddr_r
 void stm32mp_ddr_start_sw_done(struct stm32mp_ddrctl *ctl);
 void stm32mp_ddr_wait_sw_done_ack(struct stm32mp_ddrctl *ctl);
 void stm32mp_ddr_enable_axi_port(struct stm32mp_ddrctl *ctl);
+int stm32mp_ddr_disable_axi_port(struct stm32mp_ddrctl *ctl);
+void stm32mp_ddr_enable_host_interface(struct stm32mp_ddrctl *ctl);
+int stm32mp_ddr_sw_selfref_entry(struct stm32mp_ddrctl *ctl);
+void stm32mp_ddr_sw_selfref_exit(struct stm32mp_ddrctl *ctl);
 int stm32mp_board_ddr_power_init(enum ddr_type ddr_type);
 
 #endif /* STM32MP_DDR_H */
-- 
2.53.0

